extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmCdiProcessor.java

```diff
@@ -251,6 +302,47 @@ void generateDataSourceBeans(HibernateOrmRecorder recorder,
                             .createWith(recorder.statelessSessionSupplier(persistenceUnitName))
                             .addInjectionPoint(ClassType.create(DotName.createSimple(TransactionSessions.class)))
                             .done());
+
+            syntheticBeanBuildItemBuildProducer
+                    .produce(createSyntheticBean(persistenceUnitName,
+                            isDefaultPU, isNamedPU,
+                            CriteriaBuilder.class, CRITERIA_BUILDER_EXPOSED_TYPES, false)
+                            .createWith(recorder.criteriaBuilderSupplier(persistenceUnitName))
+                            .addInjectionPoint(ClassType.create(DotName.createSimple(JPAConfig.class)))
+                            .done());
+
+            syntheticBeanBuildItemBuildProducer
+                    .produce(createSyntheticBean(persistenceUnitName,
+                            isDefaultPU, isNamedPU,
+                            jakarta.persistence.metamodel.Metamodel.class, METAMODEL_EXPOSED_TYPES, false)
+                            .createWith(recorder.metamodelSupplier(persistenceUnitName))
+                            .addInjectionPoint(ClassType.create(DotName.createSimple(JPAConfig.class)))
+                            .done());
+
+            syntheticBeanBuildItemBuildProducer
+                    .produce(createSyntheticBean(persistenceUnitName,
+                            isDefaultPU, isNamedPU,
+                            SchemaManager.class, SCHEMA_MANAGER_EXPOSED_TYPES, false)
+                            .createWith(recorder.schemaManagerSupplier(persistenceUnitName))
+                            .addInjectionPoint(ClassType.create(DotName.createSimple(JPAConfig.class)))
+                            .done());
+
+            syntheticBeanBuildItemBuildProducer
+                    .produce(createSyntheticBean(persistenceUnitName,
+                            isDefaultPU, isNamedPU,
+                            Cache.class, CACHE_EXPOSED_TYPES, false)
+                            .createWith(recorder.cacheSupplier(persistenceUnitName))
+                            .addInjectionPoint(ClassType.create(DotName.createSimple(JPAConfig.class)))
+                            .done());
+
+            syntheticBeanBuildItemBuildProducer
+                    .produce(createSyntheticBean(persistenceUnitName,
+                            isDefaultPU, isNamedPU,
+                            jakarta.persistence.PersistenceUnitUtil.class, PERSISTENCE_UNIT_UTIL_EXPOSED_TYPES, false)
+                            .createWith(recorder.persistenceUnitUtilSupplier(persistenceUnitName))
+                            .addInjectionPoint(ClassType.create(DotName.createSimple(JPAConfig.class)))
+                            .done());
+
```

Comment:
Given this is becoming a lot of duplicated code, maybe you could extract it to a separate method (along with the Session/StatelessSession definition) and call the method both here and above?